File: 1249. Minimum Remove to Make Valid Parentheses/1249.py
class Solution:
    def minRemoveToMakeValid(self, s: str) -> str:
        left = list()
        right = list()

        for i, v in enumerate(s):
            if v == '(':
                left.append(i)
            elif v == ')':
                if left:
                    left.pop()
                else:
                    right.append(i)

        res = right+left
        res.sort(reverse=True)

        ns = ""
        for i, v in enumerate(s):
            if res and i == res[-1]:
                res.pop()
                continue
            ns += v

        return ns

    def minRemoveToMakeValid1(self, s: str) -> str:
        status_dict = {}
        stack = []
        # keep indexe's status in status_dict. If a parenthesis is valid
        # mark it as True else False. use stack to check it's validity
        for idx, ch in enumerate(s):
            if ch == "(":
                stack.append(idx)
            elif ch == ")" and len(stack) > 0:
                    status_dict[idx] = True
                    status_dict[stack[-1]] = True
                    stack.pop()
        res = []
        for idx, ch in enumerate(s):
            if ch == "(" or ch == ")":
                if idx in status_dict: # only append the parenthesis which are valid
                    res.append(ch)
            else:
                res.append(ch)
        return "".join(res)

a = Solution()
print(a.minRemoveToMakeValid("(a(b(c)d)"))        

File: 1249. Minimum Remove to Make Valid Parentheses/1249.go
package leetcode

import(
	"sort"
)

func minRemoveToMakeValid(s string) string {
	right := make([]int, 0)
	r := 0
	left := make([]int, 0)

	for i,v := range s{
		if v == '('{
			if r>=len(right){
				right = append(right, i)
			}else{
				if r == 0{
					right[0] = i
				}else{
					right[r] = i
				}
			}
			r++
		}else if v == ')'{
			if r>0{
				r --
			}else{
				left = append(left, i)
			}
		}
	}

	res := make([]int, len(left))
	copy(res, left)
	for i:=0; i<r; i++{
		res = append(res, right[i])
	}
	
	sort.Ints(res)
	ns := ""
	count := 0
	for i,v := range s{
		if count<len(res) && res[count] == i{
			count++
			continue
		}
		ns += string(v)
	}
	return ns
}